### PR TITLE
Validation refactoring

### PR DESCRIPTION
**Pull request by mbarnes**
### What this PR does

These are miscellaneous tweaks and refactoring commits related to both field visibility enforcement and preflight.  I factored these "fluff" commits out so that subsequent pull requests will be shorter and easier to review.

Related Jiras:

- [ARO-6347: Validate metadata from api structs](https://issues.redhat.com/browse/ARO-6347)
- [ARO-6156: Create a Preflight Endpoint for ARO HCP](https://issues.redhat.com/browse/ARO-6156)

### Special notes for your reviewer

I recommend reviewing one commit at a time.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: frontend/README.md [hunk 1]
```diff
@@ -48,7 +48,7 @@ curl -X GET "https://localhost:8443/subscriptions/YOUR_SUBSCRIPTION_ID/resourceG
 
 Create or Update a HcpOpenShiftClusterResource
 ```bash
-curl -X PUT "https://localhost:8443/subscriptions/YOUR_SUBSCRIPTION_ID/resourceGroups/YOUR_RESOURCE_GROUP_NAME/providers/Microsoft.RedHatOpenshift/hcpOpenShiftClusters/YOUR_CLUSTER_NAME?api-version=2024-06-10-preview" -H "Content-Type: application/json" -d @mycluster.yaml
```
Comment (s-amann) [inline frontend/README.md:51]: is `mycluster.json` an existing file?

Comment (mbarnes) [reply frontend/README.md:51]: No it's just an example.  I mainly wanted to fix the `.yaml` suffix as it was misleading.

File: internal/api/registry.go [hunk 2]
```diff
@@ -102,10 +109,52 @@ func NewValidator() *validator.Validate {
 type validateContext struct {
 	// Fields must be exported so valdator can access.
 	Method   string
-	Updating bool
 	Resource any
 }
 
-func ValidateRequest(validate *validator.Validate, method string, updating bool, resource any) error {
-	return validate.Struct(validateContext{Method: method, Updating: updating, Resource: resource})
+func ValidateRequest(validate *validator.Validate, method string, resource any) []arm.CloudErrorBody {
```
Comment (s-amann) [inline internal/api/registry.go:115]: wdyt about unit tests for this?

File: internal/api/v20240610preview/hcpopenshiftclusters_methods.go [hunk 3]
```diff
@@ -305,8 +294,34 @@ func (c *HcpOpenShiftClusterResource) Normalize(out *api.HCPOpenShiftCluster) {
...
+	// FIXME Validate visibility tags by comparing the new cluster (c) to current.
+
+	c.Normalize(&normalized)
+
+	errorDetails = api.ValidateRequest(validate, method, &normalized)
+	if errorDetails != nil {
+		cloudError.Details = append(cloudError.Details, errorDetails...)
+	}
+
+	switch len(cloudError.Details) {
```
Comment (s-amann) [inline internal/api/v20240610preview/hcpopenshiftclusters_methods.go:316]: this looks overly complicated, are we expecting eventually to have more than one error?

```
if len(cloudError.Details) == 0 {
  return nil
}
cloudError.CloudErrorBody = &cloudError.Details[0]
return cloudError
```

Comment (mbarnes) [reply internal/api/v20240610preview/hcpopenshiftclusters_methods.go:316]: Yes, the intent is to report _all_ validation errors in the response, not just the first one encountered.

The way this works is the logic anticipates multiple validation errors.  Multiple validation errors will look something like this:
```
{
  "Code": "MultipleErrorsOccurred",
  "Message": "Content validation failed on multiple fields",
  "Details": [
    {
      "Code": "InvalidRequestContent",
      "Message": "something about the value was bad",
      "Target": "properties.spec.the.bad.field"
    },
    ... additional errors for other fields ...
  ],
}
```

But if there's only a single validation error then the nesting isn't necessary, so it gets "promoted" into the parent section like this:
```
{
  "Code": "InvalidRequestContent",
  "Message": "something about the value was bad",
  "Target": "properties.spec.the.bad.field"
}
```

The complexity is just trying to improve readability for the end user.

**Comment by mbarnes** [comment]
Force pushed in order to rebase on https://github.com/Azure/ARO-HCP/pull/68.

File: internal/api/registry.go (hunk 2 above)
Comment (mbarnes) [reply internal/api/registry.go:115]: @s-amann I added a new commit for unit tests that cover `required_for_put` validation among other things.  If it's ok with you, I'll defer testing `ValidateRequest` to my preflight branch where the error structure in the response can be even more complicated than what I described in another comment here due to the possibility of multiple resources in the preflight request.

**Pull request merged**